fault-quarantine-module/pkg/informer/k8s_client.go

```
@@ -0,0 +1,490 @@
+	if node.Spec.Unschedulable {
+		if exist {
+			klog.Infof("Node %s already cordoned by FQM; skipping taint/annotation updates", nodename)
+			return true
+		}
+
+		klog.Infof("Node %s is cordoned manually; applying FQM taints/annotations", nodename)
+	} else {
+		klog.Infof("Cordoning node %s", nodename)
+
+		if !c.DryRunMode {
+			node.Spec.Unschedulable = true
+		}
+	}
+
+	return false
+}
+
+func (c *FaultQuarantineClient) applyAnnotations(node *v1.Node, annotations map[string]string, nodename string) {
+	if len(annotations) == 0 {
+		return
+	}
+
+	if node.Annotations == nil {
+		node.Annotations = make(map[string]string)
+	}
+
+	klog.Infof("Setting annotations %+v on node %s", annotations, nodename)
+
+	for annotationKey, annotationValue := range annotations {
+		node.Annotations[annotationKey] = annotationValue
+	}
+}
+
+func (c *FaultQuarantineClient) applyLabels(node *v1.Node, labels map[string]string, nodename string) {
+	if len(labels) == 0 {
+		return
+	}
+
+	if node.Labels == nil {
+		node.Labels = make(map[string]string)
+	}
+
+	klog.Infof("Adding labels on node %s", nodename)
+
+	for k, v := range labels {
+		node.Labels[k] = v
+	}
+}
+
+func (c *FaultQuarantineClient) UnTaintAndUnCordonNodeAndRemoveAnnotations(
+	ctx context.Context,
+	nodename string,
+	taints []config.Taint,
+	isUnCordon bool,
+	annotationKeys []string,
+	labelsToRemove []string,
+	labels map[string]string,
+) error {
+	updateFn := func(node *v1.Node) error {
+		if shouldReturn := c.removeTaints(node, taints, nodename); shouldReturn {
+			return nil
+		}
+
+		c.handleUncordon(node, isUnCordon, labels, nodename)
+
+		c.removeAnnotations(node, annotationKeys, nodename)
+
+		c.removeLabels(node, labelsToRemove, nodename)
+
+		return nil
+	}
+
+	return c.nodeInformer.UpdateNode(ctx, nodename, updateFn)
+}
+
+func (c *FaultQuarantineClient) removeTaints(node *v1.Node, taints []config.Taint, nodename string) bool {
+	if len(taints) == 0 {
```

When would the length of taints be 0? Can we perform upstream validations and fail fast?